sdk/pinocchio/src/sysvars/slot_hashes/mod.rs

```diff
@@ -0,0 +1,443 @@
+        unsafe {
+            crate::sysvars::get_sysvar_unchecked(
+                buffer.as_mut_ptr(),
+                &SLOTHASHES_ID,
+                offset,
+                buffer.len(),
+            )
+        }?;
+
+        Ok(())
+    }
+
+    /// Returns the number of `SlotHashEntry` items accessible.
+    #[inline(always)]
+    pub fn len(&self) -> usize {
+        self.len
+    }
+
+    /// Returns `true` if there are no entries.
+    #[inline(always)]
+    pub fn is_empty(&self) -> bool {
+        self.len == 0
+    }
+
+    /// Returns the entire slice of entries. Call once and reuse the slice if you
+    /// need many look-ups.
+    #[inline(always)]
+    pub fn entries(&self) -> &[SlotHashEntry] {
+        self.as_entries_slice()
+    }
+
+    /// Gets a reference to the entry at `index` or `None` if out of bounds.
+    #[inline(always)]
+    pub fn get_entry(&self, index: usize) -> Option<&SlotHashEntry> {
+        self.as_entries_slice().get(index)
+    }
+
+    /// Finds the hash for a specific slot using binary search.
+    ///
+    /// Returns the hash if the slot is found, or `None` if not found.
+    /// Assumes entries are sorted by slot in descending order.
+    /// If calling repeatedly, prefer getting `entries()` in caller
+    /// to avoid repeated slice construction.
+    #[inline(always)]
+    pub fn get_hash(&self, target_slot: Slot) -> Option<&[u8; HASH_BYTES]> {
+        let entries = self.as_entries_slice();
+        entries
+            .binary_search_by(|probe_entry| probe_entry.slot().cmp(&target_slot).reverse())
+            .ok()
+            .map(|index| &entries[index].hash)
+    }
+
+    /// Finds the position (index) of a specific slot using binary search.
+    ///
+    /// Returns the index if the slot is found, or `None` if not found.
+    /// Assumes entries are sorted by slot in descending order.
+    /// If calling repeatedly, prefer getting `entries()` in caller
+    /// to avoid repeated slice construction.
+    #[inline(always)]
+    pub fn position(&self, target_slot: Slot) -> Option<usize> {
+        let entries = self.as_entries_slice();
+        entries
+            .binary_search_by(|probe_entry| probe_entry.slot().cmp(&target_slot).reverse())
+            .ok()
+    }
+
+    /// Returns a `&[SlotHashEntry]` view into the underlying data.
+    ///
+    /// The constructor (in the safe path that called `parse_and_validate_data`)
+    /// or caller (if unsafe `new_unchecked` path) is responsible for ensuring
+    /// the slice is big enough and properly aligned.
+    #[inline(always)]
+    fn as_entries_slice(&self) -> &[SlotHashEntry] {
+        if self.len == 0 {
+            return &[];
+        }
+
+        debug_assert!(self.data.len() >= NUM_ENTRIES_SIZE + self.len * ENTRY_SIZE);
+
+        unsafe {
+            from_raw_parts(
+                self.data.as_ptr().add(NUM_ENTRIES_SIZE) as *const SlotHashEntry,
+                self.len,
+            )
+        }
```

Comment:
*minor*
Is there a reason for the `len == 0` check?
I think it would work without it as well.